garageScript/c0d3-app
# feat: Redirect to /login if user is unauthenticated

## Overview

We failed [in the last attempt to add a streamlined solution for protecting pages](https://github.com/garageScript/c0d3-app/pull/2756) because of a [bug with the next-auth package](https://github.com/nextauthjs/next-auth/issues/4234). This PR creates a hook that redirects the user to the `login` page if they're not logged in, and updates the `/settings/account` and `/review/[lessonSlug]` pages. In the future, when next-auth fixes its bug, we'll replace this temporary solution to protect the page with a [next.js middleware](https://next-auth.js.org/tutorials/securing-pages-and-api-routes#nextjs-middleware). ~~After this PR is merged, we'll release the page and share it with the students.~~

## Changes

- Create `useRedirectionUnauthenticated` hook that takes a value and redirects the user to the `/login` page if the value is true (by @JasirZaeem)
- Add a test file for `useRedirectionUnauthenticated`
- Update the `/settings/account` page test file
- Reference the new hook in both the `/settings/account` and `/review/[lessonSlug]` pages
- Create a constant for the session status and update the session status references in the declaration types file `auth` with it

## Testing

0. Logout if you're not already
1. Go to `/settings/account`
2. Notice that it shows a loading spinner and then immediately redirects you to the `/login` page
3. Login
4. Go to `/settings/account`
5. Notice it displays the page without redirecting you to `/login`

### Related issues

- #1629
